hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/MembershipNamenodeResolver.java

```diff
@@ -478,4 +478,37 @@ private List<MembershipState> getRecentRegistrationForQuery(
   public void setRouterId(String router) {
     this.routerId = router;
   }
+
+  /**
+   * Shuffle cache, to ensure that the current nn will not be accessed first next time.
+   *
+   * @param nsId name service id
+   * @param namenode namenode contexts
+   */
+  @Override
+  public synchronized void shuffleCache(String nsId, FederationNamenodeContext namenode) {
+    cacheNS.compute(Pair.of(nsId, false), (ns, namenodeContexts) -> {
+      if (namenodeContexts != null && namenodeContexts.size() > 1) {
+        FederationNamenodeContext firstNamenodeContext = namenodeContexts.get(0);
+        /*
+         * 1.If the first nn in the cache is active, the active nn priority cannot be lowered.
+         * This happens when other threads have already updated the cache.
+         *
+         * 2.If the first nn in the cache at this time is not the nn
+         * that needs to be lowered in priority, there is no need to rotate.
+         * This happens when other threads have already rotated the cache.
+         */
+        if (!firstNamenodeContext.getState().equals(ACTIVE)
+                && firstNamenodeContext.getRpcAddress().equals(namenode.getRpcAddress())) {
+          List<FederationNamenodeContext> rotatedNnContexts = new ArrayList<>(namenodeContexts);
+          Collections.rotate(rotatedNnContexts, -1);
+          return rotatedNnContexts;
+        } else {
+          return namenodeContexts;
```

Comment:
Same we can reverse the if and do early return.

Reply:
done in pr #5990